gunicorn/workers/moz_svc_gevent_worker.py

```diff
@@ -0,0 +1,238 @@
+
+
+class MozSvcGeventWorker(GeventWorker):
+    """Custom gunicorn worker with extra operational niceties.
+
+    This is a custom gunicorn worker class, based on the standard gevent worker
+    but with some extra operational- and debugging-related features:
+
+        * a background thread that monitors execution by checking for:
+
+            * blocking of the gevent event-loop, with tracebacks
+              logged if blocking code is found.
+
+            * overall memory usage, with forced-gc and graceful shutdown
+              if memory usage goes beyond a defined limit.
+
+        * a timeout enforced on each individual request, rather than on
+          inactivity of the worker as a whole.
+
+        * a signal handler to dump memory usage data on SIGUSR2.
+
+    To detect eventloop blocking, the worker installs a greenlet trace
+    function that increments a counter on each context switch.  A background
+    (os-level) thread monitors this counter and prints a traceback if it has
+    not changed within a configurable number of seconds.
+    """
+
+    def init_process(self):
+        # Check if we need a background thread to monitor memory use.
+        needs_monitoring_thread = False
+        if MAX_MEMORY_USAGE:
+            self._last_memory_check_time = time.time()
+            needs_monitoring_thread = True
+
+        # Set up a greenlet tracing hook to monitor for event-loop blockage,
+        # but only if monitoring is both possible and required.
+        if hasattr(greenlet, "settrace") and MAX_BLOCKING_TIME > 0:
+            # Grab a reference to the gevent hub.
+            # It is needed in a background thread, but is only visible from
+            # the main thread, so we need to store an explicit reference to it.
+            self._active_hub = gevent.hub.get_hub()
+            # Set up a trace function to record each greenlet switch.
+            self._active_greenlet = None
+            self._greenlet_switch_counter = 0
+            greenlet.settrace(self._greenlet_switch_tracer)
+            self._main_thread_id = _real_get_ident()
+            needs_monitoring_thread = True
+
+        # Create a real thread to monitor out execution.
+        # Since this will be a long-running daemon thread, it's OK to
+        # fire-and-forget using the low-level start_new_thread function.
+        if needs_monitoring_thread:
+            _real_start_new_thread(self._process_monitoring_thread, ())
+
+        # Continue to superclass initialization logic.
+        # Note that this runs the main loop and never returns.
+        super(MozSvcGeventWorker, self).init_process()
+
+    def init_signals(self):
+        # Leave all signals defined by the superclass in place.
+        super(MozSvcGeventWorker, self).init_signals()
+
+        # Hook up SIGUSR2 to dump memory usage information.
+        # This will be useful for debugging memory leaks and the like.
+        signal.signal(signal.SIGUSR2, self._dump_memory_usage)
```

Comment:
I agree with @wgallagher that scoping this change to just handle pause detection makes the most sense.